File: src/JavaScript/StructuralPatterns/Composite/WindowsFormsApplication/composites/baseControlComposite.js

const Control = require('../component/control');

class BaseControlComposite extends Control {
    constructor(name, width, height, foreColor, backColor) {
        super(name, width, height, foreColor, backColor);
        this.controls = new Set();
    }

    add(control) {
        this.controls.add(control);
    }

    remove(control) {
        this.controls.delete(control);
    }

    getHierarchicalLevel(depthIndicator) {
        let groupBoxHierarchicalInfo = `${'-'.repeat(depthIndicator)}+ ` +
            `Name: ${this.name}, Width: ${this.width}, Fore Color: (${this.foreColor.red}, ${this.foreColor.green}, ${this.foreColor.blue}), ` +
            `Back Color: (${this.backColor.red}, ${this.backColor.green}, ${this.backColor.blue})\r\n`;

        for (const control of this.controls) {
            groupBoxHierarchicalInfo += control.getHierarchicalLevel(depthIndicator + 2);
        }

        return groupBoxHierarchicalInfo;
    }
}

module.exports = BaseControlComposite;

File: src/Java/StructuralPatterns/Composite/WindowsFormsApplication/src/Composites/BaseControlComposite.java
package Composites;

import Component.Control;

import java.awt.Color;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public abstract class BaseControlComposite extends Control {
    private List<Control> controls = new ArrayList<>();

    public BaseControlComposite(String name, int width, int height, Color foreColor, Color backColor) {
        super(name, width, height, foreColor, backColor);
    }

    @Override
    public List<Control> getControls() {
        return controls;
    }

    @Override
    public void add(Control control) {
        controls.add(control);
    }

    @Override
    public void remove(Control control) {
        controls.remove(control);
    }

    @Override
    public String getHierarchicalLevel(int depthIndicator) {
        StringBuilder compositeHierarchicalInfo = new StringBuilder(
                String.format("%s+ Name: %s, Width: %d, Fore Color: (%d, %d, %d), Back Color: (%d, %d, %d)\r\n",
                        new String(new char[depthIndicator]).replace('\0', '-'),
                        getName(), getWidth(),
                        getForeColor().getRed(), getForeColor().getGreen(), getForeColor().getBlue(),
                        getBackColor().getRed(), getBackColor().getGreen(), getBackColor().getBlue()
                )
        );

        for (Control control : controls) {
            compositeHierarchicalInfo.append(control.getHierarchicalLevel(depthIndicator + 2));
        }

        return compositeHierarchicalInfo.toString();
    }
}
